README.md

```diff
@@ -11,47 +11,62 @@
+
+This entry-point sets up a static 16 kilobyte stack before invoking
+`__sel4_start_root()`, also provided by the runtime, which
+constructs the argument, environment, and auxiliary vectors. It then
+passes the constructed vectors, along with the address of `main()`,
+into `__sel4_start_main` which configures the runtime according to the
+architecture's conventions before starting calling `main()`.
+
+### Standard Processes
+
+All processes other than the root task use the `_start` entry-point provided in
 the architecture-dependant `crt0.S`. This will then bootstrap into the
 runtime entry-point `__sel4_start_c` which simply processes the stack to
 find the argument, environment, and auxiliary vectors.
 
-The found vectors, along with`main`, are passed into
+These vectors, along with the address of `main()`, are passed into
 `__sel4_start_main` which configures the runtime before starting
-`main`.
+`main`.  The architecture's standard post-main() processing is handled
+by `__start`.
 
-## Root Task
 
-The root task requires an alternate entry-point `_sel4_start` which
-assumes that the `seL4_BootInfo` argument has been passed to it and that
-it has not been given a stack.
+### Thread-local storage layout
 
-This entry-point moves onto a static 16 kilobyte stack before invoking
-`__sel4_start_root`, which constructs the argument, environment, and
-auxiliary vectors. It then passes the constructed vectors, along with
-`main`, into `__sel4_start_main` which configures the runtime before
-starting `main`.
+There are two supported layouts for thread local storage.
 
-## Thread-local storage layout
+Intel X86\_64 and IA32 use a TLS pointer that contains the address
+immediately after the TLS region: the region is _below_ it in memory.
+Other architectures use a TLS pointer that contains the lowest address in
+the TLS.
 
-There are two standard layouts for thread local storage commonly used.
-One where the TLS base address refers to the first address in memory of
-the region and one where it refers to the address that immediately
-follows the region. Intel's x86_64 and ia32 architectures use the latter
-method as it aligns with the segmentation view of memory presented by
-the processor. Most other platforms use former method, where the TLS can
-be said to be 'above' the thread pointer.
+For X86, memory immediately _above_ the address is available for the C
+library to use as thread metadata.  For other architectures, two words are
+allocated right at the top of the TLS region.  The first word is
+initialised to point to the thread local storage, either at one past
+its highest address (X86 family) or at its lowest address (other
+architectures). 
 
-In order to store metadata for the current thread in the same memory
-allocation as the TLS, the run-time utilises memory on the other side of
-the thread pointer for it's thread structure.
+In any case, the size of the TLS is determined by values in the ELF
+file from which the program is loaded.
```

Comment:
It may be worth mentioning that in order for the runtime to determine the size of the TLS in a root task, the ELF file needs to be linked with specifically named symbols located at the start and end of the TLS region. For other applications, the runtime is passed the size and location of the TLS on the stack from whatever loads the image into memory.

Reply:
Yes, I found it really confusing to work out where this info came from.

Reply:
One thing I noticed recently, is that many of our binaries that we load include the program headers in the first page of the first PT_LOAD segment which would indicate that the root task would be able to find the TLS segment itself and we wouldn't need to have all of the additional linker scripts we have.  I'm not sure how standard this is though, it may just be the current version of binutils and x86_64 linker script I currently have.

Reply:
I personally want to make the boot process set up a System-V stack for the root task so that no dedicated runtime is required for root tasks.